File: tools/bench_read.py
"""Measure what reading a ``.KeyStepPro`` project costs, phase by phase.

Evidence for requirement D1 -- "a parser that reads them in a timely,
byte-efficient and non-duplicative manner". Times the bytes off disk, the
lenient JSON parse and the decode to the model separately, and reports peak
memory as a ratio to the file's own size on disk.

    uv run python tools/bench_read.py project_files/project_5.KeyStepPro
    uv run python tools/bench_read.py --json --reps 9 project_files/baseline.KeyStepPro
    uv run python tools/bench_read.py --render < readings.jsonl

In ``tools/`` rather than ``src/``: a dev instrument, not a shipped command.
"""

from __future__ import annotations

import argparse
import json
import platform
import resource
import statistics
import sys
import time
import tracemalloc
from collections.abc import Iterator, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any

# Allow running from a source checkout without installing the package.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "src"))

from ksp import lenient_json, reader

# ru_maxrss is bytes on Darwin and kilobytes everywhere else.
RSS_UNIT = 1 if sys.platform == "darwin" else 1024

TRACKS_PER_PROJECT = 4

# One `lru_cache` hit is below `perf_counter`'s resolution, so it is timed in a batch.
REPEAT_CALLS = 1000

# Column key -> heading, in the order the table prints them.
COLUMNS = {
    "read_bytes_s": "bytes",
    "parse_json_s": "json",
    "decode_s": "decode",
    "total_s": "total",
    "repeat_load_s": "repeat",
}


@dataclass(frozen=True)
class Rep:
    """One uncached read, phase by phase, plus a second ``load`` of the same path."""

    read_bytes_s: float
    parse_json_s: float
    decode_s: float
    repeat_load_s: float

    @property
    def total_s(self) -> float:
        return self.read_bytes_s + self.parse_json_s + self.decode_s


def one_rep(path: Path) -> Rep:
    """Time one read with the ``lru_cache`` cleared, then one repeat that hits it."""
    reader.load.cache_clear()

    start = time.perf_counter()
    data = path.read_bytes()
    after_bytes = time.perf_counter()
    raw = lenient_json.loads(data)
    after_json = time.perf_counter()
    project = reader.read_project(raw, source_name=path.name)
    after_decode = time.perf_counter()

    repeated = reader.load(path)
    before_repeat = time.perf_counter()
    for _ in range(REPEAT_CALLS):
        reader.load(path)
    after_repeat = time.perf_counter()
    if repeated.device != project.device:
        raise RuntimeError(f"{path.name}: the repeated load disagrees with the decoded project")

    return Rep(
        read_bytes_s=after_bytes - start,
        parse_json_s=after_json - after_bytes,
        decode_s=after_decode - after_json,
        repeat_load_s=(after_repeat - before_repeat) / REPEAT_CALLS,
    )


def traced_memory(path: Path) -> dict[str, int]:
    """What one read holds and where its high-water mark falls, by ``tracemalloc``.
    Runs after the reps, so the model figure excludes ``keys.key``'s one-off 4096-entry cache."""
    reader.load.cache_clear()
    tracemalloc.start()
    try:
        data = path.read_bytes()
        held_bytes, _ = tracemalloc.get_traced_memory()

        tracemalloc.reset_peak()
        raw = lenient_json.loads(data)
        held_raw, parse_peak = tracemalloc.get_traced_memory()

        tracemalloc.reset_peak()
        project = reader.read_project(raw, source_name=path.name)
        held_decoded, decode_peak = tracemalloc.get_traced_memory()

        del data, raw
        held_model, _ = tracemalloc.get_traced_memory()
    finally:
        tracemalloc.stop()

    if len(project.tracks) != TRACKS_PER_PROJECT:
        raise RuntimeError(f"{path.name}: decoded {len(project.tracks)} tracks, expected 4")

    return {
        "traced_peak_bytes": max(parse_peak, decode_peak),
        "traced_held_bytes_bytes": held_bytes,
        "traced_held_raw_bytes": held_raw,
        "traced_held_decoded_bytes": held_decoded,
        "traced_model_bytes": held_model,
    }


def rss_bytes() -> int:
    return int(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss) * RSS_UNIT


def format_ms(seconds: float) -> str:
    """Milliseconds, kept to three significant figures down to the cache hit's 50 ns."""
    value = seconds * 1e3
    if value >= 1:
        return f"{value:.2f}"
    return f"{value:.4f}" if value >= 0.001 else f"{value:.6f}"


def spread(values: Sequence[float]) -> dict[str, float]:
    """Min is the honest floor; median is what a person waits on."""
    return {"min": min(values), "median": statistics.median(values)}


def measure(path: Path, reps: int, floor_bytes: int) -> dict[str, Any]:
    one_rep(path)  # discarded warm-up
    samples = [one_rep(path) for _ in range(reps)]
    traced = traced_memory(path)
    size = path.stat().st_size
    peak = rss_bytes()

    reading: dict[str, Any] = {
        "core": "python",
        "file": path.name,
        "size_bytes": size,
        "reps": reps,
    }
    for key in COLUMNS:
        reading[key] = spread([float(getattr(sample, key)) for sample in samples])
    reading |= traced
    reading |= {
        "traced_ratio": traced["traced_peak_bytes"] / size,
        "traced_model_ratio": traced["traced_model_bytes"] / size,
        "rss_peak_bytes": peak,
        "rss_ratio": (peak - floor_bytes) / size,
        "rss_floor_bytes": floor_bytes,
        "runtime": f"CPython {platform.python_version()}",
        "platform": f"{platform.system()} {platform.release()} {platform.machine()}",
    }
    return reading


def human(reading: dict[str, Any]) -> Iterator[str]:
    """One reading, one phase per line."""
    size = int(reading["size_bytes"])
    yield f"{reading['file']}  ({size / 1e6:.2f} MB, {reading['reps']} reps)"
    for key, heading in COLUMNS.items():
        figures = reading[key]
        yield (
            f"  {heading:<7} min {format_ms(figures['min']):>10} ms"
            f"   median {format_ms(figures['median']):>10} ms"
        )
    for label, key in (
        ("the file's bytes", "traced_held_bytes_bytes"),
        ("after the parse", "traced_held_raw_bytes"),
        ("after the decode", "traced_held_decoded_bytes"),
        ("the model alone", "traced_model_bytes"),
        ("high-water mark", "traced_peak_bytes"),
    ):
        held = int(reading[key])
        yield f"  {label:<17}{held / 1e6:8.2f} MB   = {held / size:6.2f} x the file"
    yield (
        f"  rss peak    {int(reading['rss_peak_bytes']) / 1e6:7.2f} MB"
        f"   = {reading['rss_ratio']:.2f} over a"
        f" {int(reading['rss_floor_bytes']) / 1e6:.2f} MB runtime floor"
    )


def render(readings: Sequence[dict[str, Any]]) -> Iterator[str]:
    """Both cores side by side, medians in milliseconds, one row per core per file."""
    head = "".join(f"{heading:>11}" for heading in COLUMNS.values())
    yield f"{'file':<28}{'size':>8}  {'core':<7}{head}{'held/byte':>11}{'rss/byte':>10}"
    yield "-" * (28 + 8 + 2 + 7 + len(head) + 21)

    seen: list[str] = []
    for name in [str(reading["file"]) for reading in readings]:
        if name not in seen:
            seen.append(name)

    for name in seen:
        first = True
        for reading in readings:
            if reading["file"] != name:
                continue
            size = int(reading["size_bytes"])
            label = f"{name:<28}{size / 1e6:7.2f}M" if first else " " * 36
            first = False
            times = "".join(f"{format_ms(reading[key]['median']):>11}" for key in COLUMNS)
            traced = reading.get("traced_ratio")
            traced_cell = "--" if traced is None else f"{traced:.2f}"
            yield (
                f"{label}  {reading['core']!s:<7}{times}"
                f"{traced_cell:>11}{reading['rss_ratio']:>10.2f}"
            )
    yield ""
    yield "Times are medians in milliseconds. held/byte is the tracemalloc peak per byte of file;"
    yield "rss/byte is the process peak above the runtime floor, per byte of file."


def main(argv: list[str] | None = None) -> int:
    floor_bytes = rss_bytes()

    parser = argparse.ArgumentParser(
        description="Measure what reading one .KeyStepPro project costs, phase by phase.",
    )
    parser.add_argument("file", nargs="?", type=Path, help="one project file")
    parser.add_argument("--reps", type=int, default=5, help="timed reps after one warm-up")
    parser.add_argument("--json", action="store_true", help="one machine-readable line")
    parser.add_argument(
        "--render",
        action="store_true",
        help="read --json lines from stdin and print the combined table instead",
    )
    args = parser.parse_args(argv)

    if args.render:
        readings = [json.loads(line) for line in sys.stdin if line.strip()]
        for line in render(readings):
            print(line)
        return 0

    if args.file is None:
        parser.error("a project file is required without --render")
    if args.reps < 1:
        parser.error("--reps must be at least 1")

    reading = measure(args.file, args.reps, floor_bytes)
    if args.json:
        print(json.dumps(reading))
    else:
        for line in human(reading):
            print(line)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/bench_read.swift
// Measure what reading a `.KeyStepPro` project costs, phase by phase: requirement D1's evidence.
//
// Compiled against KSPKit by scripts/bench_read.sh, which runs it once per sample so that a peak
// memory figure belongs to one file. `<file> [--reps N] [--json]`, or `--render` over --json lines.

import Foundation

// `@main` rather than top-level code: swiftc allows that only in a file called `main.swift`.
@main
enum BenchRead {
    /// One `ReadCache` hit is below the clock's resolution, so it is timed in a batch.
    static let repeatCalls = 1000

    static let columns: [(key: String, heading: String)] = [
        ("read_bytes_s", "bytes"),
        ("parse_json_s", "json"),
        ("decode_s", "decode"),
        ("total_s", "total"),
        ("repeat_load_s", "repeat"),
    ]

    struct BenchError: Error, CustomStringConvertible {
        let description: String
    }

    static func seconds(_ duration: Duration) -> Double {
        let (whole, attoseconds) = duration.components
        return Double(whole) + Double(attoseconds) / 1e18
    }

    /// One uncached read, phase by phase, then the repeat `load` that hits `ReadCache`.
    static func oneRep(_ url: URL) throws -> [String: Double] {
        Reader.clearCache()
        let clock = ContinuousClock()
        let start = clock.now
        let data = try Data(contentsOf: url)
        let afterBytes = clock.now
        let raw = try LenientJSON.parse(data)
        let afterJSON = clock.now
        let project = try Reader.readProject(raw, sourceName: url.lastPathComponent)
        let afterDecode = clock.now

        let repeated = try Reader.load(contentsOf: url)
        let beforeRepeat = clock.now
        for _ in 0..<repeatCalls {
            _ = try Reader.load(contentsOf: url)
        }
        let afterRepeat = clock.now
        guard repeated == project, project.tracks.count == 4 else {
            throw BenchError(description: "\(url.lastPathComponent): the repeated load disagrees")
        }

        let bytes = seconds(afterBytes - start)
        let json = seconds(afterJSON - afterBytes)
        let decode = seconds(afterDecode - afterJSON)
        return [
            "read_bytes_s": bytes,
            "parse_json_s": json,
            "decode_s": decode,
            "total_s": bytes + json + decode,
            "repeat_load_s": seconds(afterRepeat - beforeRepeat) / Double(repeatCalls),
        ]
    }

    // ru_maxrss is bytes on Darwin, kilobytes on Linux.
    static func rssBytes() -> Int {
        var usage = rusage()
        getrusage(RUSAGE_SELF, &usage)
        return Int(usage.ru_maxrss)
    }

    static func median(_ values: [Double]) -> Double {
        let sorted = values.sorted()
        let middle = sorted.count / 2
        return sorted.count % 2 == 1 ? sorted[middle] : (sorted[middle - 1] + sorted[middle]) / 2
    }

    static func measure(_ url: URL, reps: Int, floor: Int) throws -> [String: Any] {
        _ = try oneRep(url)  // discarded warm-up
        let samples = try (0..<reps).map { _ in try oneRep(url) }
        let size = try FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int ?? 0
        let peak = rssBytes()

        var reading: [String: Any] = [
            "core": "swift",
            "file": url.lastPathComponent,
            "size_bytes": size,
            "reps": reps,
            "rss_peak_bytes": peak,
            "rss_floor_bytes": floor,
            "rss_ratio": Double(peak - floor) / Double(size),
            "runtime": "swiftc -O",
        ]
        for column in columns {
            let values = samples.map { $0[column.key] ?? 0 }
            reading[column.key] = ["min": values.min() ?? 0, "median": median(values)]
        }
        return reading
    }

    /// Milliseconds, kept to three significant figures down to the cache hit's 50 ns.
    static func milliseconds(_ seconds: Double) -> String {
        let value = seconds * 1e3
        if value >= 1 { return String(format: "%.2f", value) }
        return String(format: value >= 0.001 ? "%.4f" : "%.6f", value)
    }

    static func padded(_ text: String, _ width: Int, left: Bool = false) -> String {
        let fill = String(repeating: " ", count: max(0, width - text.count))
        return left ? text + fill : fill + text
    }

    static func phase(_ reading: [String: Any], _ key: String, _ figure: String) -> Double {
        (reading[key] as? [String: Double])?[figure] ?? 0
    }

    static func human(_ reading: [String: Any]) -> [String] {
        let size = Double(reading["size_bytes"] as? Int ?? 0)
        var lines = [
            "\(reading["file"] ?? "")  (\(String(format: "%.2f", size / 1e6)) MB, "
                + "\(reading["reps"] ?? 0) reps)"
        ]
        for column in columns {
            lines.append(
                "  \(padded(column.heading, 7, left: true)) min "
                    + "\(padded(milliseconds(phase(reading, column.key, "min")), 10)) ms   median "
                    + "\(padded(milliseconds(phase(reading, column.key, "median")), 10)) ms")
        }
        let peak = Double(reading["rss_peak_bytes"] as? Int ?? 0) / 1e6
        let floor = Double(reading["rss_floor_bytes"] as? Int ?? 0) / 1e6
        lines.append(
            "  rss peak    \(String(format: "%7.2f", peak)) MB   = "
                + "\(String(format: "%.2f", reading["rss_ratio"] as? Double ?? 0)) over a "
                + "\(String(format: "%.2f", floor)) MB runtime floor")
        return lines
    }

    /// Medians in milliseconds, one row per file.
    static func render(_ readings: [[String: Any]]) -> [String] {
        let head = columns.map { padded($0.heading, 11) }.joined()
        var lines = [
            padded("file", 28, left: true) + padded("size", 8) + head + padded("rss/byte", 10),
            String(repeating: "-", count: 28 + 8 + head.count + 10),
        ]
        for reading in readings {
            let size = Double(reading["size_bytes"] as? Int ?? 0) / 1e6
            let times = columns.map { padded(milliseconds(phase(reading, $0.key, "median")), 11) }
            lines.append(
                padded(reading["file"] as? String ?? "", 28, left: true)
                    + padded(String(format: "%.2fM", size), 8) + times.joined()
                    + padded(String(format: "%.2f", reading["rss_ratio"] as? Double ?? 0), 10))
        }
        lines.append("")
        lines.append("Times are medians in milliseconds. rss/byte is the process peak above the")
        lines.append("runtime floor, per byte of file.")
        return lines
    }

    static func usage() -> Never {
        FileHandle.standardError.write(
            Data("usage: bench_read <file> [--reps N] [--json] | bench_read --render\n".utf8))
        exit(2)
    }

    static func main() throws {
        let floor = rssBytes()
        var arguments = Array(CommandLine.arguments.dropFirst())

        if arguments == ["--render"] {
            let input = String(
                decoding: FileHandle.standardInput.readDataToEndOfFile(), as: UTF8.self)
            let readings = try input.split(separator: "\n").compactMap {
                try JSONSerialization.jsonObject(with: Data($0.utf8)) as? [String: Any]
            }
            for line in render(readings) {
                print(line)
            }
            return
        }

        var reps = 5
        var json = false
        if let flag = arguments.firstIndex(of: "--json") {
            json = true
            arguments.remove(at: flag)
        }
        if let flag = arguments.firstIndex(of: "--reps") {
            guard flag + 1 < arguments.count, let count = Int(arguments[flag + 1]), count >= 1
            else { usage() }
            reps = count
            arguments.removeSubrange(flag...(flag + 1))
        }
        guard arguments.count == 1 else { usage() }

        let reading = try measure(URL(filePath: arguments[0]), reps: reps, floor: floor)
        if json {
            let line = try JSONSerialization.data(withJSONObject: reading, options: [.sortedKeys])
            print(String(decoding: line, as: UTF8.self))
        } else {
            for line in human(reading) {
                print(line)
            }
        }
    }
}
